### PR TITLE
Add existing sniffs to validate all input, escape all output.

### LINKED ISSUE #37: [Implement sniff] Validate and/or sanitize all input, escape all output
#### Rule type:
Error

#### Rule:

> Validate and/or sanitize untrusted data before entering into the database. All untrusted data should be escaped before output.

Ref: https://make.wordpress.org/themes/handbook/review/required/#code

#### Theme check file covering this rule:
N/A

#### To do:
- [ ] Add existing `WordPress.XSS.EscapeOutput` sniff to the ruleset.
- [ ] Add existing `WordPress.VIP.ValidatedSanitizedInput` sniff to the ruleset.

### PR DESCRIPTION
**Pull request by jrfnl**
Check for validation/sanitization of all input, escaping of all output.

Fixes #37

### CODE REVIEW

File: WordPress-Theme/ruleset.xml [hunk 1]
```diff
@@ -21,9 +21,12 @@
 	<!-- No ByteOrderMark allowed - important to prevent issues with content being sent before headers. -->
 	<rule ref="Generic.Files.ByteOrderMark" />
 
-
 	<!-- Check that the I18N functions are used correctly. -->
 	<!-- This sniff can also check the text domain, provided it's passed to PHPCS. -->
 	<rule ref="WordPress.WP.I18n" />
 
+	<!-- Validate and/or sanitize untrusted data before entering into the database.
```
Comment (grappler) [inline WordPress-Theme/ruleset.xml]: I don't think the sniff checks if the data is sanitized, just that it is escaped.

**Comment by grappler** [comment]
The sniff `WordPress.XSS.EscapeOutput` checks if the i18n functions are escaped which we do not require.

I think we will need to change the notice from ERROR to WARNING on non WordPress core functions and variables. The reason being there may be no need for escaping.

I don't think we should allow the ignore comment `//xss ok`

File: WordPress-Theme/ruleset.xml (hunk 1 above)
Comment (jrfnl) [inline WordPress-Theme/ruleset.xml]: You're right, I'd forgotten to add the other sniff in haste. Corrected now.

**Comment by jrfnl** [comment]
@grappler 

> The sniff WordPress.XSS.EscapeOutput checks if the i18n functions are escaped which we do not require.

The rule in the handbook doesn't explicitly exclude i18n output. So in that case the handbook would need adjusting.

Also - just putting this out there:
Is it justified to exclude i18n output ? I know that GlotPress translations are (very?) often used and  - with the changes in WP 4.6 - will take precedence. As those translations are reviewed, I can see how they could be regarded as save, but themes can still ship their own translations, can't they ?

> I think we will need to change the notice from ERROR to WARNING on non WordPress core functions and variables. The reason being there may be no need for escaping.

> I don't think we should allow the ignore comment //xss ok

The escaping sniff has two-fold additional functionality to deal with this:

1. People can provide an comma delimited list of `customEscapingFunctions`, `customAutoEscapedFunctions` and `customPrintingFunctions` via the properties of their `phpcs.xml` file to whitelist additional functions.
2. The `// xss ok` whitelist comment.

Considering the future Theme Check will disregard a custom `phpcs.xml` file as we don't want rules to be disabled, that leaves authors only one other option - which is the whitelist comment.

As both the sanitation and the escaping is IMHO **_very_** important for the application security, lowering the error level to warning feels counter-intuitive.

What about leaving the level at `error` and allowing the whitelist comments, *but* also adding an additional sniff for those same whitelist comments which would output a `warning` saying the variable will need manual inspection ?

**Comment by grappler** [comment]
> Is it justified to exclude i18n output ? 

The translation are reviewed on w.org by translation editor who would not approve a translation if there were issues with it. Core does not use the escaped functiosn Ref: https://github.com/WordPress/twentysixteen/pull/240 & https://core.trac.wordpress.org/ticket/30724

My hesitation for allowing the `// xss ok` whitelist comment is the additional thing to understand and if it will be used as a lazy workaround. Need to get feedback from others on this. The biggest issue that we face is with `get_theme_mod()`.


As to the validation/sanitization sniff 
- There is a few false positive that I need to report upstream e.g.
 - `wp_verify_nonce( $_POST[ 'settings_nonce' ], basename( __FILE__ )`
 - `if ( 'page' === $_POST['post_type'] ) {`
 - https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/191
- It requires the data to be run through `wp_unslash()` before it is sanitized which I don't fully understand.

Otherwise it is good. 😄

**Comment by jrfnl** [comment]
> The translation are reviewed on w.org by translation editor who would not approve a translation if there were issues with it.

Only for translations made in GlotPress, not for translations shipped with a theme. Or is that no longer allowed at all ?

> Core does not use the escaped functiosn 

Core does not comply with quite a few of their own coding standard directives - largely because they only allow code to be adjusted to comply with the standard in combination with other changes. So IMHO, that's not really a valid argument.

As for TwentySixteen - I'd call that "leading by setting a bad example".

> My hesitation for allowing the // xss ok whitelist comment is the additional thing to understand and if it will be used as a lazy workaround. Need to get feedback from others on this.

Should we change the label for the issue to `decision needed` in that case ?

> The biggest issue that we face is with get_theme_mod().

Can you explain that one to me ?


> There is a few false positive that I need to report upstream

Please do, they should be solved upstream.

> wp_verify_nonce( $_POST[ 'settings_nonce' ], basename( __FILE__ )

Already reported - see the response in the issue: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/572

> It requires the data to be run through wp_unslash() before it is sanitized which I don't fully understand.

Core slashes all input data, so to do a valid comparison, you'd need to unslash the data before doing the compare.

Hope this helps.

**Comment by grappler** [comment]
> Only for translations made in GlotPress, not for translations shipped with a theme. Or is that no longer allowed at all ?

Translations are still allowed as w.org only ships translations when they are at 100%. There may be easier way for theme authors to slip in malicious code if they wanted to. I have added the decision needed tag.

> The biggest issue that we face is with get_theme_mod().

I just meant that the sniff is catching places where the settings are not being escaped. It is a common thing the theme authors miss.

**Comment by jrfnl** [comment]
> I just meant that the sniff is catching places where the settings are not being escaped. It is a common thing the theme authors miss.

So that's a good thing, isn't it ?

**Comment by grappler** [comment]
> So that's a good thing, isn't it ?

Yes, meant it as a good thing.

**Comment by justintadlock** [comment]
A decision was reached on the idea of escaping translations and having to add `// xss ok` to bypass escaping variables.  The team voted down both proposals.

Reference: https://make.wordpress.org/themes/2016/10/04/october-4th-team-meeting-notes/

Anything that wouldn't block submission (warning, info) would probably be acceptable.

**Comment by jrfnl** [comment]
I've lowered the error level to `warning` for both sniffs in accordance with the board's decision.

For the record:
While I respect the decision by the theme review board, this seems a very counter-intuitive direction to take.
Both these sniffs are about security and are now given low priority while this is one of the biggest issues with both themes and plugins for WP.

Very few "false positives" for these sniffs which have been reported to WPCS upstream actually turn out to be false positives. Most of the time, it is developer education which is needed, not an adjustment of the sniff.

I strongly suggest to put it on the agenda for reconsideration in six months or so once the Theme ruleset has shaped up more and more full theme checks have been run.

If you do encounter false positives in the mean time, please *do* report these upstream.

**Pull request merged**